### PR TITLE
Visualizations

### PR DESCRIPTION
**Pull request by sylwiabr**
Creating custom visualizations documentation based on plotting library example

### CODE REVIEW

File: visualizations.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
+#Visualizations in Luna
+
+Visualization in Luna is a widget responsible for displaying data in certain way in Node Editor. In Luna Studio we have few types of visualizations available for different types of data, i.e. echarts plots for lists.
+To show the visualziation you have to click the eye icon on the right side of node name:
+![](/assets/visualizations_eye_icon.png)
+To choose the different type of visualization you can drop a list of possible to use types of visualizations.
+![](/assets/visualizations_drop_list.png)
```
Comment (lgawlik) [inline visualizations.md]: I don't understand what is this picture form, what is shown on it

File: visualizations.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+```
+  window.addEventListener("message", function (evt) {
+    if(evt.data.data) {
+        data = JSON.parse(evt.data.data);
+        dataToPlot = data;
+        ExamplePlottingLibrary.plotCommand("example_div", dataToPlot)
+    };
+    }
+```
+If you want the visualization behave in certain way on "resize" or "load" message you can define it in the same file in proper event listeners (look at the template - https://github.com/luna/visualizers-template/blob/master/example/example.js).
```
Comment (ljkania) [inline visualizations.md]: `resize` or `load`

File: visualizations.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        };
+
+    if (cfgHelper.matchesType(type, examplePattern))
+        {
+            return [{name: "example", path: "example.html"}]
+        }
+    else
+        return [];
+};
+```
```
Comment (ljkania) [inline visualizations.md]: It is worth commenting that return value is list of all visualizers served by this packet

File: visualizations.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    }
+```
+If you want the visualization behave in certain way on "resize" or "load" message you can define it in the same file in proper event listeners (look at the template - https://github.com/luna/visualizers-template/blob/master/example/example.js).
+
+To set on which type visualization should be shown we need to add a constructor to constructors list in `config.js` file:
+```
+module.exports = function (type) {
+    var examplePattern =
+        { constructor: ["Plot"], fields: [{constructor: ["Int", "Real"], fields: { any: true }}]
+        };
```
Comment (ljkania) [inline visualizations.md]: Isn't there any better layout?

File: visualizations.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+```
+If a node will have `Plot` type the `example` visualization will be shown in drop-down menu on the left side of the node.
+
+To use such defined visualization in your project you need to define the list of Int which will be your data:
+![](/assets/visualizations_list.png)
+and a new node packing it in the `DataSeries` type:
+![](/assets/visualizations_dataseries.png)
+and then into `Plot` type with single element list contains your data:
+![](/assets/visualizations_plot.png)
+
```
Comment (ljkania) [inline visualizations.md]: Something is not rendered as intended, or at least it looks like it

File: visualizations.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        Main.luna
+    visualizers/
+        example/
+            config.js
+            example.css
+            example.html
+            example.js            
+```
+It contains few files:
+    - `min.js` - minified JS file with library you want to connect with visualization (not included in visualizer template)
```
[the remark this answers is not in the document]
Comment (lgawlik) [reply visualizations.md]: In Firefox browse, the bullet list (min.js, config.js, ...) are not correctly formated, all the points are displayed as one sentence, without line breaks

File: visualizations.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+```
+class PlotType:
+    Bar
+    Pie
+    Scatter
+
+class DataSeries:
+    data:: List Int
+    type:: PlotType
+```
```
Comment (lgawlik) [inline visualizations.md]: In which file I should add these classes?

File: visualizations.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+Data series can be an Int list and the plot type can be `Scatter` , `Pie` or `Bar`. I will create separate type called `PlotType`. Additional helpers for `DataSeries` are defined to make creating JSON more simple and straightforward. To use it in JS we need to define `toJSON` method on each of the classes.
```
Comment (lgawlik) [inline visualizations.md]: PlotType class was written above, what is the meaning of sentence "I will create separate type called `PlotType`" here?

File: visualizations.md [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+class Plot:
+    Plot
+    PlotVal (List DataSeries)
+    def toJSON: case self of
+        PlotVal ds: ds.toJSON
+```
+
+To read the data in JS in `example.js` file you need to unpack the JSON:
+```
+  window.addEventListener("message", function (evt) {
```
Comment (lgawlik) [inline visualizations.md]: In downloaded example.js there is not such a function, downloaded example should be exactly what we are describing in the post

File: visualizations.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+    if (cfgHelper.matchesType(type, examplePattern))
+        {
+            return [{name: "example", path: "example.html"}]
+        }
+    else
+        return [];
+};
+```
+If a node will have `Plot` type the `example` visualization will be shown in drop-down menu on the left side of the node.
```
Comment (lgawlik) [inline visualizations.md]: this drop-down menu is not visible on the picture below, I think it would be good to show it

File: visualizations.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
+## Visualizations in Luna
+
+Visualization in Luna is a widget responsible for displaying data in certain way in Node Editor. In Luna Studio we have few types of visualizations available for different types of data, i.e. echarts plots for lists.
+To show the visualziation you have to click the eye icon on the right side of node name:
+![](/assets/visualizations_eye_icon.png)
```
Comment (ljkania) [inline visualizations.md]: This icon is not shown properly

File: visualizations.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+![](/assets/visualizations_list.png)
+and a new node packing it in the `DataSeries` type:
+![](/assets/visualizations_dataseries.png)
+and then into `Plot` type with single element list contains your data:
+![](/assets/visualizations_plot.png)
+
+And voila! you can see your plot in Node Editor!
+
+To see more visualizations implementation please visit:
+https://github.com/luna-packages/PlotLu
```
Comment (lgawlik) [inline visualizations.md]: Link does not work, 404 error

**Pull request closed without merging**